### PR TITLE
Feature/routing

### PR DESCRIPTION
**Pull request by mmcachran**
Adds routing to Tribe Libs.

### CODE REVIEW

**Comment by defunctl** [comment]
Hey @mmcachran,

Looks like all tests are failing due to invalid JSON in the composer.lock, can you please get the test passing?

Thanks!

**Review by nickpelton — COMMENTED** [review]
Going to try and give this a test in this next week.

File: src/Routes/Abstract_REST_Route.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
+<?php // phpcs:disable WordPress.Files.FileName.NotHyphenatedLowercase
```
Comment (defunctl) [inline src/Routes/Abstract_REST_Route.php:1]: I think we should actually name this `Abstract_Rest_Route` and make this work on case sensitive file systems: https://github.com/moderntribe/square-one/pull/678/files#r606454579

File: src/Routes/Abstract_REST_Route.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	 */
+	abstract public function register() : void;
+
+	/**
+	 * Returns the WP API namespace.
+	 *
+	 * @return string The WP API namespace.
+	 */
+	public function get_api_namespace() : string {
+		return 'wp/v2';
```
Comment (defunctl) [inline src/Routes/Abstract_REST_Route.php:38]: How often would this change across an entire project do you think?

File: src/Routes/Abstract_Route.php [hunk 3]
```diff
@@ -0,0 +1,356 @@
+<?php // phpcs:disable WordPress.Files.FileName.NotHyphenatedLowercase
```
Comment (defunctl) [inline src/Routes/Abstract_Route.php:1]: Don't think this is needed in this class

File: src/Routes/Abstract_Route.php [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+/**
+ * Base class for routes.
+ */
+abstract class Abstract_Route {
+	/**
+	 * Registers any hooks for the route.
+	 *
+	 * @return void
+	 */
+	abstract public function register() : void;
```
Comment (defunctl) [inline src/Routes/Abstract_Route.php:52]: As mentioned in the other PR, we should not register any hooks in our instances, so probably won't need this.

File: src/Routes/Abstract_Route.php [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+
+	/**
+	 * Required - The Route's name. This is useful to refer to the route
+	 * by name. Eg:- \MTTrial\is_route( 'foo' );
+	 *
+	 * @return string The name for the route.
+	 */
+	public function get_name() : string {
```
Comment (defunctl) [inline src/Routes/Abstract_Route.php:60]: Same deal here for `:` spacing, please.

File: src/Routes/Abstract_Route.php [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+
+	/**
+	 * The request methods that are authorized on this route. Only GET
+	 * is authorized by default.
+	 *
+	 * @return array Accepted request methods for the route.
+	 */
+	public function get_request_methods() : array {
+		return [
+			'GET',
```
Comment (defunctl) [inline src/Routes/Abstract_Route.php:214]: Should use the `WP_REST_Server` constants here as well

File: src/Routes/Abstract_Route.php [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+		];
+	}
+
+	/**
+	 * Starts the Route's lifecyle hooks by activating the route.
+	 *
+	 * @param \WP $wp The main WordPress object.
+	 * @return void
+	 */
+	public function activate( $wp ) : void {
```
Comment (defunctl) [inline src/Routes/Abstract_Route.php:224]: Let's chat about moving this into a subscriber instead.

File: src/Routes/Abstract_Route.php [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+		if ( empty( $template_path ) ) {
+			return $template;
+		}
+
+		global $wp_query;
+
+		// Bail early if this is the 404 template.
+		if ( strpos( $template_path, '404' ) !== false ) {
+			$wp_query->is_404 = true;
+			$protocol         = filter_input( INPUT_SERVER, 'SERVER_PROTOCOL', FILTER_SANITIZE_SPECIAL_CHARS );
```
Comment (defunctl) [inline src/Routes/Abstract_Route.php:285]: Doesn't this not work on all PHP configurations, e.g. those using fastcgi?

File: src/Routes/Abstract_Route.php [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+		if ( defined( 'PHPUNIT_RUNNER' ) ) {
+			return $valid;
+		}
+
+		// Bail early if the method is valid.
+		if ( $valid ) {
+			return $valid;
+		}
+
+		header( 'HTTP/1.1 403 Forbidden' );
```
Comment (defunctl) [inline src/Routes/Abstract_Route.php:353]: We don't actually know which HTTP version a server could be using.

File: src/Routes/Abstract_Route.php [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+		wp_die( 'Not Authorized' );
```
Comment (defunctl) [inline src/Routes/Abstract_Route.php:354]:
```suggestion
		wp_die( 'Not Authorized', 403 );
```
Wouldn't this work properly instead?

File: src/Routes/Router.php [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+
+namespace Tribe\Libs\Routes;
+
+use Tribe\Libs\Container\Abstract_Subscriber;
+use Tribe\Libs\Routes\Abstract_Route;
+
+/**
+ * Class to register routers for normal and REST API endpoints.
+ */
+class Router extends Abstract_Subscriber {
```
Comment (defunctl) [inline src/Routes/Router.php:20]: Ah, I see now that you're extending the Container's subscriber for this. I'm not sure this is a very good idea, in the sense that responsibility for a subscriber should be to register instances to WordPress hooks. This should really be its own stand alone/testable class that is instead registered for the proper hooks in the Router_Subscriber.

**Review by defunctl — CHANGES_REQUESTED** [review]
Some things to chat about here, but looks like a great place to start.

In general, please add `@action` and `@filter` comments to your methods if they are hooked into WordPress. Makes it easier for other developers to see that, e.g. `@action init` or `@filter the_content` etc....

Would very much like to see some tests for this, just reach out if need some direction.

**Comment by defunctl** [comment]
@mmcachran this failed PHPCS and shouldn't have been merged (sorry we didn't have this set up to prevent that): https://github.com/moderntribe/tribe-libs/runs/2401691633

**Comment by defunctl** [comment]
@mmcachran and @nickpelton  A few other issues here:

1. https://github.com/moderntribe/square1-routes/ should be created via: https://github.com/moderntribe/tribe-libs#adding-packages
1. Then it should be added to https://github.com/moderntribe/tribe-libs/blob/master/monorepo-builder.php#L38
1. It should then get registered to packagist.
1. I realize the automated workflow bumped the version, but given this is new functionality following [semver](https://semver.org/) it should have been bumped to `3.5.0`. Ideally, rolling back the releases and releasing this with the proper version, and they have to be manually removed from packagist.

**Pull request merged**